experimental/oauthtokenretriever/token.go

```diff
@@ -0,0 +1,128 @@
+package oauthtokenretriever
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+type tokenRetriever struct {
+	GrafanaAppURL string
+	ClientID      string
+	ClientSecret  string
+	HTTPClient    *http.Client
+
+	signer signer
+}
+
+// tokenPayload returns a JWT payload for the given user ID, client ID, and host.
+func (t *tokenRetriever) tokenPayload(userID string) map[string]interface{} {
+	iat := time.Now().Unix()
+	exp := iat + 1800
+	u := uuid.New()
+	payload := map[string]interface{}{
+		"iss": t.ClientID,
+		"sub": fmt.Sprintf("user:id:%s", userID),
+		"aud": t.GrafanaAppURL + "/oauth2/token",
+		"exp": exp,
+		"iat": iat,
+		"jti": u.String(),
+	}
+	return payload
+}
+
+type jwtBearer struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int    `json:"expires_in"`
+	Scope       string `json:"scope"`
+	TokenType   string `json:"token_type"`
+}
+
+// retrieveJWTBearerToken returns a JWT bearer token for the given user ID.
+func (t *tokenRetriever) retrieveJWTBearerToken(userID string) (string, error) {
```

Comment:
in this case, I was not able to use the `jwt` package since I couldn't find a way to set the necessary `client_id` and `client_secret` (the endpoint returned an error because that info was missing).

Reply:
another reason: that library only supports RS256 keys, in our case the default is ES256:

https://github.com/golang/oauth2/blob/master/internal/oauth2.go#L20

Reply:
Thanks for looking into this :heart: